### ISSUE TITLE
Protocols should generate sealed class hierarchies

### ISSUE THREAD

**Issue by julianpeeters** [julianpeeters/avrohugger#32]
Currently protocol support is paltry, and simply generates separate files for the classes defined therein, whereas Java codegen generates an interface.

Seems natural to represent the interface as a sealed trait, and generate the classes in a single compilation unit.

**Comment by julianpeeters** [comment]
This is almost complete, and much functionality can be found in release version 0.11.0. What remains to do is the reflective compilation of ADTs into ADTs whose members implement SpecificRecordBase or Scavro's AvroSerializable.

**Comment by mariussoutier** [comment]
Is this expected behavior that a single record will not create a sealed class hierarchy?

**Comment by julianpeeters** [comment]
Yes, but only because of I'm unaware of a use case for single-member hierarchies, and was worried that some people might consider it untidy, extraneous cruft.

A reason to generate traits even for single-record protocols would be great, one less complication to document.

**Comment by mariussoutier** [comment]
The reason in this case is that the protocol is still work in progress and we kicked out some unused records, so for the time being only one record is left. We were quite surprised that the remaining record no longer inherited from the protocol trait and had to add a dummy record to bring the hierarchy back.

**Comment by julianpeeters** [comment]
Whoops, didn't see that one coming, but a pretty good reason IMO :)

Ok, so what remains to be done here is:
1. Scala to Scala code gen
2. Restore support for single-member hierarchies

**Comment by francescopellegrini** [comment]
HI, any update on this? :)
